### DISCUSSION TITLE
How to build cloudsimplus source and cloudsimplus-examples together

### DISCUSSION THREAD

**Discussion by deRemo** [manoelcampos/cloudsim-plus#420] [Q&A] [answered]
Hi all,

I'm trying to merge the two maven projects (cloudsimplus and cloudsimplus-examples) so that I can skim through the simulator source code while trying the examples using the search functionality of VSCode - much like I would do with original cloudsim, since both the simulator and the examples are bundled under the same gradle/maven project. 

However, I have zero experience with such building tools, can you give me a help? I tried creating a new folder containing both source codes (simulator and examples) and merging the related maven projects via a multi-module root pom.xml, but it didn't work 

Thank you

**Comment by manoelcampos** [comment, marked as answer]
Hello @deRemo 

You can create a multi-module project by introducing a new pom.xml file inside the root folder where you place both API and examples project. The [release 7.2.0 has exactly this file you can copy](https://github.com/cloudsimplus/cloudsimplus/blob/v7.2.0/pom.xml). 

If you use IntelliJ, you can just open the root folder that you can navigate and build both projects.

But keep in mind that you don't need to change the framework code to create your simulations. You're expected to just extend existing classes if you need to implement some new feature. If you are changing the framework code, you're probably doing something wrong. If you do that, you endup with a custom version of the framework that will be difficult to update when new official versions are released.

**Comment by deRemo** [comment]
You are right, but I'm no maven expert and I would like to work on the internals for bug-fixing.
For whoever may be interested, that's how I configured my workspace:

1) clone cloudsimplus repository and build it
2) clone cloudsimplus-examples and change the cloudsimplus dependency in the root pom.xml to:
```
        <dependency>
            <groupId>org.cloudsimplus</groupId>
            <artifactId>cloudsimplus</artifactId>
            <version>${project.version}</version>
            <scope>system</scope>
            <systemPath>/path/to/cloudsimplus/target/cloudsimplus-8.5.1.jar</systemPath>
        </dependency>
```
3) Make your changes in cloudsimplus repository and compile
4) Make your changes in cloudsimplus-example repository and run example

**Reply by manoelcampos** [reply]
If you change the cloudsimplus-examples pom.xml you may end up committing these changes. If you send a PR with those changes, they won't work for other users since it's using a local file path.

Copying the file I mentioned before may work out-of-the-box. But if you need some help, let me know.

**Reply by deRemo** [reply]
Yes you are right, you have to be extra-careful with what you push (but it would be the same with the multi-module file, I believe?). Anyway, I'll try it the "right" way if I'll get the chance, thanks

**Reply by manoelcampos** [reply]
You can add the the new pom.xml to .gitignore

**Reply by manoelcampos** [reply]
In fact, since the new pom.xml file will be outside of the individual projects' directories, it won't be tracked and ignored by default.

**Reply by deRemo** [reply]
I was convinced that I was supposed to change the current pom.xml in cloudsimplus-example, I misread! Today I'll give this a try, thanks for the clarifications

**Comment by deRemo** [comment]
Hi @manoelcampos, I used the pom.xml that you linked me, changed the module names (cloudsim-plus -> cloudsimplus) and removed the ones I don't need. However, cloudsimplus-examples is still tied to the maven's version of cloudsimplus, it seems like I necessarily have to make the changes I was saying above, unless I'm missing something.

**Reply by deRemo** [reply]
I was indeed missing something, the module name for my local cloudsimplus' sources was different from the corresponding folder name. I had some problem with linking my local sources with the examples folder (for example, when I used the goto declaration functionality on Intellij, it would go to the decompiled target, not the actual source code in the submodule). I genuinely don't know how I fixed that, I simply kept re-building dependencies, cleaning caches etc :(

**Reply by manoelcampos** [reply]
I’ll check that. On IntelliJ it usually works right away

**Comment by deRemo** [comment]
To make sure that Intellij jumps to the source code, instead of the compiled classes:

1) File > Project Structure > Libraries
2) Remove the Sources path of the compiled version
3) Add /path/to/cloudsimplus/src/main/java

**Comment by jlxcsong** [comment]
您好！您发的邮件我已收到！
